core/src/providers/textsynth.rs

```diff
@@ -116,12 +116,166 @@ impl TextSynthLLM {
-    //     Ok(format!("https://api.textsynth.com/v1/engines/{}/chat", self.id).parse::<Uri>()?)
-    // }
+    fn chat_uri(&self) -> Result<Uri> {
+        Ok(format!("https://api.textsynth.com/v1/engines/{}/chat", self.id).parse::<Uri>()?)
+    }
+
+    // This function is in charge of formatting the messages for the chat TextSynth interface by
+    // injecting the role and name as context to the model. As a result the model will echo that
+    // structure. We use `extract_name_and_role` to clean it up.
+    fn format_chat_messages(&self, messages: &Vec<ChatMessage>) -> (Option<String>, Vec<String>) {
+        let mut system: Option<String> = None;
+        let mut messages = messages.clone();
+
+        // If the first message is a system message, remove it and return a system string.
+        if messages.len() > 0 && messages[0].role == ChatMessageRole::System {
+            system = match messages[0].content {
+                Some(ref c) => Some(c.clone()),
+                None => None,
+            };
+            messages.remove(0);
+        }
+
+        let formatted = messages
+            .iter()
+            .map(|m| {
+                let content = match m.content {
+                    Some(ref c) => c.clone(),
+                    None => String::from(""),
+                };
+
+                match m.name {
+                    Some(ref n) => format!("role={} name={} {}", m.role.to_string(), n, content),
+                    None => format!("role={} {}", m.role.to_string(), content),
+                }
+            })
+            .collect::<Vec<_>>();
+
+        (system, formatted)
+    }
+
+    // Cleans up output of the model by removing the injected role and name pieces. We run this in
+    // non streamed mode on the final response. In streamed mode we run it on the first chunk
+    // (textsynth returned more than one token at a time and this will fall into it with high
+    // likelihood; this is not perfect but usage will tell if we need to be more elaboratde).
+    fn extract_name_and_role(text: &str) -> (Option<ChatMessageRole>, Option<String>, String) {
+        let mut response_text = text.trim().to_string();
+        let mut role: Option<ChatMessageRole> = None;
+        let mut name: Option<String> = None;
+
+        // Extract and remove `role=...` at the beginning of the message if it exists.
+        let re = regex::Regex::new(r"^role=(\w+) ").unwrap();
+        response_text = match re.captures(response_text.as_str()) {
+            Some(c) => {
+                let r = c.get(1).unwrap().as_str().to_string();
+                match r.as_str() {
+                    "assistant" => role = Some(ChatMessageRole::Assistant),
+                    "user" => role = Some(ChatMessageRole::User),
+                    "system" => role = Some(ChatMessageRole::System),
+                    _ => (),
+                }
+                response_text[6 + r.len()..].to_string().trim().to_string()
```

Comment:
Question: instead of hardcoding a start at 6, why not replacing the string "role=" by empty string? I'm not fluent in Rust but something like that

if let Some(c) = RE.captures(response_text.as_str()) {
    let matched = c.get(0).unwrap().as_str();
    response_text = response_text.replacen(matched, "", 1).trim().to_string();
    // ...
}

Reply:
That's quite equivalent no? I really don't like unwraping so I have a tendency to shy from it :p

Reply:
Will leave it as is out of laziness :grimacing: